sdk/core/Azure.Core/src/Shared/GZipUtf8JsonRequestContent.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System.IO;
+using System.IO.Compression;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Core
+{
+    internal class GZipUtf8JsonRequestContent: RequestContent
+    {
+#pragma warning disable CA2213 // Disposable fields should be disposed
+        private GZipStream _gzip;
+        private MemoryStream _stream;
+#pragma warning restore CA2213 // Disposable fields should be disposed
+        private readonly RequestContent _content;
+
+        public Utf8JsonWriter JsonWriter { get; }
+
+        public GZipUtf8JsonRequestContent(RequestContent content)
+        {
+            _stream = new MemoryStream();
+            _gzip = new GZipStream(_stream, CompressionMode.Compress, true);
+            _content = Create(_stream);
```

Comment:
It seems like we are doing too many copies with this implementation. This create method creates StreamRequestContent, which stores a stream and then it's written to the network, it reads the stored stream to an array and then writes it to the destination. I think this filed (_content) should be simply MemoryStream and not RequestContent, and then we would avoid some of the unnecessary copies.

Reply:
The copying is primarily to read the uncompressed input `RequestContext` into the newly created `RequestContext` (`_context`) which is required to compress it into the gzip stream.  I'm not sure we can avoid these copies to accomplish this. But I could be missing something.

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/30691